openshift/console
# ODC-7210: Create Serverless function using the Import from the Git with Builder Images

**Fixes**: 
https://issues.redhat.com/browse/ODC-7210

**Solution Description:**

Added **Serverless Function** Import Strategy to the Git Import Form. Users now can create Serverless Functions directly from the Git Import Form if the Git Repository contains func.yaml file.

**ACs Completed**: 

- [x] Detect func.yaml in Git repo when user enter the Git URL
- [x] Read runtime, builder, buildEnvs and run envs value from the func.yaml. func.yaml
- [x] Proceed with this flow if the builder value is s2i
- [x] Allow users to view and possibly edit the environment variables which are shown in func.yaml. Add the buildEnvs and run envs to the resource 
- [x] if func.yaml is detected change the strategy to Serverless function and do not show builder images in to and  provide a section which allows the user to select the runtime version
- [x] Do not show the Pipeline and Resources section. Default to Serverless Deployment
- [x] Add label function.knative.dev: 'true' to KSVC to visualize the Serverless function in the Topology


**Screen shots / Gifs for design review**: 


https://user-images.githubusercontent.com/47265560/209876416-422b671f-e0ad-4cd7-b35f-701af4ff83fe.mp4

**Screenshot when the New strategy gets selected**

![screencapture-localhost-9000-import-ns-avik-2023-01-09-19_58_01](https://user-images.githubusercontent.com/47265560/211331880-8672e9e7-0b1b-4dc2-80a3-aee4a89250ec.png)

**Screenshot when a Unsupported Runtime mentioned in the Func.yaml**

![screencapture-localhost-9000-import-ns-avik-2023-01-16-18_09_55](https://user-images.githubusercontent.com/47265560/212680849-2c0bc8a1-4bee-4c57-b642-10622f3f0953.png)

**Screenshot when a random Runtime mentioned in the Func.yaml**

![screencapture-localhost-9000-import-ns-avik-2023-01-16-18_10_15](https://user-images.githubusercontent.com/47265560/212680895-7329afe1-0aa4-4109-96f2-6238d1b4f8b3.png)


**With all Strategy Tiles Opened**

![screencapture-localhost-9000-import-ns-avik-2023-01-09-20_07_06](https://user-images.githubusercontent.com/47265560/211333525-5b9f83e2-2210-45f9-9866-b3f558bd3a16.png)


**Unit test coverage report**: 

E2E tests added.
Unit tests fixed.

**Test setup:**

1. Install the Serverless Operator
2. Go to the Git Import Add Form.
3. Paste a Git URL having `func.yaml` file. (e.g. https://github.com/Lucifergene/oc-func)
4. You will see the Serverless Function Strategy get selected and the Runtime and version are visible.

**Test Files:**
- `func.yaml` with `runtime` set as `node`: https://bitbucket.org/avikkundu/oc-func
-  `func.yaml` with `runtime` set as `go`: https://bitbucket.org/avikkundu/oc-func-go
-  `func.yaml` with `runtime` set as `god`: https://bitbucket.org/avikkundu/oc-func-god

**Browser conformance**:
